lib/src/authentication/credential.dart

```diff
@@ -0,0 +1,42 @@
+import 'package:http/http.dart' as http;
+import 'package:meta/meta.dart';
+
+import 'bearer.dart';
+
+typedef CredentialDeserializer = Credential Function(Map<String, dynamic>);
+
+final Map<String, CredentialDeserializer> _supportedMethods = {
+  'Bearer': BearerCredential.fromMap,
+};
+
+abstract class Credential {
+  static Credential? fromMap(Map<String, dynamic> map) {
+    final authMethod = map['method'] as String?;
+    final credentials = map['credentials'] as Map<String, dynamic>?;
+
+    if (credentials != null &&
+        authMethod?.isNotEmpty == true &&
+        _supportedMethods.containsKey(authMethod)) {
+      return _supportedMethods[authMethod]!(credentials);
+    }
+
+    return null;
+  }
+
+  Map<String, dynamic> toMap() {
+    return <String, dynamic>{
+      'method': authenticationType,
+      'credentials': toMapInternal(),
+    };
+  }
+
+  /// Authentication type of this credential.
+  String get authenticationType;
+
+  /// Add required details for authentication to [request].
```

Comment:
We can. But this will limit some possibilities like modifying custom request headers for new credential types in future (I don't know what that might be, but it's there just in case we need it).

If you think that's unnecessary I'll modify it to something like `Future<String> getAuthenticationHeader()` which just returns header value. (Future because some credentials might resolve header value asynchronously, eg: by requesting from special endpoint that generates unique token for that request: like OAuth2 token refresh) .

Reply:
There is a strong argument to be made that authentication should never be done by means other than an `Authentication` header.

I can see the argument that you might want to have access to the HTTP method and URL being requested... If the authentication mechanism is signing a JWT or HMAC (like AWS signatures v4).

But attaching other headers or using a custom `x-key` header is a very bad idea, because proxies and other tooling is often aware of the `authentication` header, and won't write these to logs, error reports, etc.